src/quilla/plugins/base_storage.py

```diff
@@ -0,0 +1,188 @@
+    def is_enabled(self) -> bool:
+        '''
+        A property to be used to determine if the plugin is configured and
+        therefore able to run
+
+        Returns:
+            True if the plugin is configured, False otherwise
+        '''
+
+    @abstractmethod
+    def find_image_by_baseline(self, baseline_id: str) -> bytes:
+        '''
+        Searches the defined storage method for the image matching some
+        baseline ID
+
+        Args:
+            baseline_id: The unique ID to search for. It is assumed every
+                baseline ID is unique regardless of what test is requesting it.
+
+        Returns:
+            The bytes representation of the stored image if found, or an empty
+            bytes object if the image is not found.
+        '''
+
+    @abstractmethod
+    def cleanup_reports(self):
+        '''
+        Searches for reports that match some cleanup criteria
+        '''
+
+    @abstractmethod
+    def store_treatment_image(
+        self,
+        run_id: str,
+        baseline_id: str,
+        treatment: bytes,
+    ) -> str:
+        '''
+        Stores a treatment image within the storage mechanism enabled by the plugin
+
+        Args:
+            run_id: The run ID of the current Quilla run, to version the treatment images
+            baseline_id: The ID of the baseline that this treatment image is associated
+                with
+            treatment: The image data in bytes
+
+        Returns:
+            An identifier that can locate the newly stored treatment image
+        '''
+
+    @abstractmethod
+    def store_baseline_image(
+        self,
+        baseline_id: str,
+        baseline: bytes,
+    ) -> str:
+        '''
+        Stores a baseline image under the given baseline_id.
+
+        This function should be used to update the current baseline
+        image, or create a new one if the baseline did not previously exist
+
+        Args:
+            baseline_id: A unique identifier for the image
+            baseline: The image data in bytes
+
+        Returns:
+            A URI for the new baseline image
+        '''
+
+    @abstractmethod
+    def make_baseline_uri(
+        self,
+        run_id: str,
+        baseline_id: str
+    ) -> str:
+        '''
+        Generates a baseline URI for the current run given the baseline_id of the image.
+
+        It is recommended that plugins create a clone of the baseline image
+        when generating the URI so that the returned URI will uniquely identify
+        the baseline that was used for the associated run ID. This ensures that
+        even if the baseline image is updated, the report is still meaningful.
+
+        Args:
+            run_id: The unique ID identifying the current run
+            baseline_id: The unique identifier for the image
+
+        Returns:
+            A URI that can locate the baseline image used for the given run
+        '''
+
+    def get_image(self, baseline_id: str) -> Optional[bytes]:
```

Comment:
how is it different from find_image_by_baseline()?

Reply:
So I’ve used this method and some of the ones below to “standardize” the way the plugins are expected to work in a way that prevents some possible bugs down the line. Namely, instead of instructing plugins to check if they are enabled before trying to resolve the image, the plug-in would only need to implement the image search function and this class would handle the logic required for the check. I’m trying to see if i can make method plug-in hooks work properly, cuz if i can just set those to be the hooks themselves that simplifies the work down the line